[DOC-12145] Added missing query to example 3 of WITH RECURSIVE page - Server

julia-cb opened the pull request:
Jira ticket: [DOC-12145](https://issues.couchbase.com/browse/DOC-12145)
Capella doc PR: #201 
[Preview doc page](https://preview.docs-test.couchbase.com/fixwithrecursive/server/current/n1ql/n1ql-language-reference/with-recursive.html) (preview site password [here](https://couchbasecloud.atlassian.net/wiki/spaces/DOCS/pages/1971224949/Docs+Team+demo+site+passwords))

Changes in this PR:
- Added new query and CREATE INDEX to example 3 of [WITH RECURSIVE page in Server](https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/with-recursive.html) (will copy query over to the capella branch once this PR is approved)
- Added new query handling hierarchical/tree-structured data from the [Recursive CTE google doc](https://docs.google.com/document/d/18ospWJSP3-P-SqI5LcMBR2z7tTQhEbmyLN3kfEQ4RtQ/edit#heading=h.adfsg822ty1r) which I hadn't added to the page before

julia-cb committed:
Added missing query to example 3

julia-cb committed:
Added "examples handling hierarchical or tree-structured data"

GauravJayaraj commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
Hi,
Thanks for the quick fix:)
While we are at it I spotted another change required here https://github.com/couchbaselabs/docs-devex/pull/196/files#diff-edbf6e825c4644d098ef7d8b929f0f2c4097b659d5a5c49a4003b296ba30fcb5R368

OPTIONS {"levels:2"} -> OPTIONS {"levels":2}

Other that this all good.

julia-cb replied:
Thanks, Gaurav! Just made the small change. I'll merge this once Simon has taken a look and reviewed it :)

julia-cb committed:
Small updates

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
The Recursive SELECT syntax diagram starting at line 43 needs a `[source,ebnf]` block attribute list

simon-dew replied:
The examples all get the word "Example" prepended to them automatically, so they don't need to contain the word "Example" 😄 

- "Example of simple recursive referencing" (line 143) should be just "Simple recursive referencing"
- "Example combining recursive and non-recursive CTEs with the WITH clause" (line 174) should be just "Combine recursive and non-recursive CTEs with the WITH clause"

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
 ----
 ====
 
+.Examples handling hierarchical or tree-structured data
```
As above, doesn't need to include the word "Example"

```suggestion
.Handle hierarchical or tree-structured data
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
 
+.Examples handling hierarchical or tree-structured data
+====
+Before trying out these examples, you must:
```
This won't work until you've created the `employees` collection. I'm not 100% sure of the best place to create this. It doesn't seem to belong in the `inventory` scope, so let's put it in `tenant_agent_00`.

Best practice in the SQL++ reference is to set the query context to the bucket and scope you'll be using for the query.

```suggestion
For this example, set the query context to the `tenant_agent_00` scope in the travel sample dataset.
For more information, see xref:n1ql:n1ql-intro/queriesandresults.adoc#query-context[Query Context].
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
+====
+Before trying out these examples, you must:
+
+. Create a dataset containing employee information:
++
```
Using numbered steps with code blocks in this example creates an inconsistency, because some some code blocks have titles and some don't, and the different indents make the code blocks look wonky. Use titled code blocks throughout the example.

```suggestion
.Create a dataset containing employee information
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
+. Create a dataset containing employee information:
++
+[source,sqlpp]
+----
```
We need to create the collection first.

```suggestion
----
CREATE COLLECTION employees IF NOT EXISTS;

```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
+RETURNING *;
+----
++
+. Create a new index:
++
```
```suggestion
.Create an index for the employees
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
+CREATE INDEX e_idx ON employees(manager_id INCLUDE MISSING, employee_id, employee_name);
+----
+
+.Example finding the level of each employee in the organization
```
```suggestion
.Find the level of each employee in the organization
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
+   ON e1.manager_id = emplLevel.employee_id
+)
+SELECT * FROM emplLevel;
+----
```
Show the expected results. (Feel free to cut the results down with an ellipsis `// ...` if you feel they're too long, as long as the different employee levels are clear.)

```suggestion
----

.Results
[source,json]
----
[
  {
    "emplLevel": {
      "employee_id": 1,
      "employee_name": "John",
      "lvl": 0
    }
  },
  {
    "emplLevel": {
      "employee_id": 2,
      "employee_name": "Emily",
      "lvl": 1,
      "manager_id": 1
    }
  },
  {
    "emplLevel": {
      "employee_id": 3,
      "employee_name": "Mike",
      "lvl": 1,
      "manager_id": 1
    }
  },
  {
    "emplLevel": {
      "employee_id": 4,
      "employee_name": "Sarah",
      "lvl": 2,
      "manager_id": 2
    }
  },
  {
    "emplLevel": {
      "employee_id": 6,
      "employee_name": "Lisa",
      "lvl": 2,
      "manager_id": 2
    }
  },
  {
    "emplLevel": {
      "employee_id": 5,
      "employee_name": "Alex",
      "lvl": 2,
      "manager_id": 3
    }
  }
]
----
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
+SELECT * FROM emplLevel;
+----
+
+.Example resulting in the reporting hierarchy of the employees in the organization
```
```suggestion
.Find the reporting hierarchy of the employees in the organization
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
+    SELECT cte.* FROM cte
+) as reportsTo
+FROM employees e;
+----
```
Show the results. (Feel free to cut the results down with an ellipsis `// ...` if you feel they're too long, as long as the different reporting structures are clear.)

```suggestion
----

.Results
[source,json]
----
[
  {
    "employee_id": 1,
    "employee_name": "John",
    "reportsTo": [
      {
        "employee_id": 2,
        "employee_name": "Emily",
        "manager_id": 1
      },
      {
        "employee_id": 3,
        "employee_name": "Mike",
        "manager_id": 1
      },
      {
        "employee_id": 4,
        "employee_name": "Sarah",
        "manager_id": 2
      },
      {
        "employee_id": 6,
        "employee_name": "Lisa",
        "manager_id": 2
      },
      {
        "employee_id": 5,
        "employee_name": "Alex",
        "manager_id": 3
      }
    ]
  },
  {
    "employee_id": 2,
    "employee_name": "Emily",
    "manager_id": 1,
    "reportsTo": [
      {
        "employee_id": 4,
        "employee_name": "Sarah",
        "manager_id": 2
      },
      {
        "employee_id": 6,
        "employee_name": "Lisa",
        "manager_id": 2
      }
    ]
  },
  {
    "employee_id": 3,
    "employee_name": "Mike",
    "manager_id": 1,
    "reportsTo": [
      {
        "employee_id": 5,
        "employee_name": "Alex",
        "manager_id": 3
      }
    ]
  },
  {
    "employee_id": 4,
    "employee_name": "Sarah",
    "manager_id": 2,
    "reportsTo": []
  },
  {
    "employee_id": 6,
    "employee_name": "Lisa",
    "manager_id": 2,
    "reportsTo": []
  },
  {
    "employee_id": 5,
    "employee_name": "Alex",
    "manager_id": 3,
    "reportsTo": []
  }
]
----
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
+----
+====
+
 .Examples using the OPTIONS clause and the CYCLE clause to avoid infinite recursion
```
```suggestion
.Use the OPTIONS clause and the CYCLE clause to avoid infinite recursion
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
+
 .Examples using the OPTIONS clause and the CYCLE clause to avoid infinite recursion
 ====
+Before trying out these examples, you must:
```
Again, let's set the query context to the bucket and scope you'll be using for the example.

```suggestion
For this example, set the query context to the `tenant_agent_00` scope in the travel sample dataset.
For more information, see xref:n1ql:n1ql-intro/queriesandresults.adoc#query-context[Query Context].
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
 ====
+Before trying out these examples, you must:
+
+. Create a dataset containing product information:
++
```
Just titles for the code blocks, as above

```suggestion
.Create a dataset containing product information
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
+. Create a dataset containing product information:
++
+[source,sqlpp]
+----
```
Create the collection first:

```suggestion
----
CREATE COLLECTION products IF NOT EXISTS;

```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
+RETURNING *;
+----
++
+. Create a new index:
++
```
```suggestion
. Create an index for the products
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -212,8 +212,140 @@ SELECT * FROM rcte;
+CREATE INDEX p_idx ON products(id); 
+----
+
 .Example resulting in infinite recursion
```
```suggestion
.Query resulting in infinite recursion
```

Also, add a line _after_ this query stating what users will see if they try to run it. Something like

```asciidoc
This query will continue running without giving any results, until Couchbase Server runs out of memory.
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -235,7 +367,7 @@ WITH RECURSIVE similar_items AS (
-OPTIONS {"levels:2"}
+OPTIONS {"levels":2}
 SELECT * FROM similar_items;
 ---- 
```
Show the expected results.

Also, add a line _after_ the query with the CYCLE clause, stating the results are the same as the previous query. You don't need to show the results twice.

```suggestion
----

.Results
[source,json]
----
[
  {
    "similar_items": {
      "category": "Sports",
      "description": "A sturdy and reliable bicycle.",
      "id": 1,
      "lvl": 0,
      "name": "Bicycle",
      "price": 299.99,
      "related_items": [
        2,
        4
      ]
    }
  },
  {
    "similar_items": {
      "category": "Sports",
      "description": "A safety helmet for cycling.",
      "id": 2,
      "lvl": 1,
      "name": "Helmet",
      "price": 49.99,
      "related_items": [
        1,
        3
      ]
    }
  },
  {
    "similar_items": {
      "category": "Sports",
      "description": "Comfortable gloves for cycling.",
      "id": 4,
      "lvl": 1,
      "name": "Gloves",
      "price": 19.99,
      "related_items": [
        1,
        3
      ]
    }
  },
  {
    "similar_items": {
      "category": "Accessories",
      "description": "A convenient water bottle for cycling.",
      "id": 3,
      "lvl": 2,
      "name": "Water Bottle",
      "price": 9.99,
      "related_items": [
        2,
        4
      ]
    }
  },
  {
    "similar_items": {
      "category": "Sports",
      "description": "A sturdy and reliable bicycle.",
      "id": 1,
      "lvl": 2,
      "name": "Bicycle",
      "price": 299.99,
      "related_items": [
        2,
        4
      ]
    }
  }
]
----
```

simon-dew replied:
Another thing I've just noticed. In the reporting hierarchy example, I think the `reportsTo` alias is misleading. The query doesn't show who each employee reports to, it shows each employee's reports, i.e. the people they manage. Please can you change `reportsTo` to `reports` in this query, and in the results?

(Also, let's change some of the employee names to add some more diversity!)

simon-dew reviewed (changes requested):
Thanks @julia-cb ! Some changes required to set the query context for the examples, create the required collections, and show the results.

julia-cb committed:
Apply suggestions from code review

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

julia-cb committed:
Apply suggestions from code review

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

julia-cb committed:
Removed travel-sample example from Server (#200)

julia-cb committed:
Updated index example in example 4, added more excample results

julia-cb commented:
@simon-dew Thanks for all the feedback!! Super helpful and all great suggestions, I applied all of them :) (also totally agree on the name diversity for the employee example - I updated it with different names)

I updated the preview site to reflect all the changes you requested! https://preview.docs-test.couchbase.com/fixwithrecursive/server/current/n1ql/n1ql-language-reference/with-recursive.html

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -226,6 +501,17 @@ WITH RECURSIVE similar_items AS (
 SELECT * FROM similar_items;
 ----
 
+.Results
```
This is fine as-is, but for this one example I think it would also be okay just to say "this query will run until the Query service runs out of memory".

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/with-recursive.adoc:
```diff
@@ -251,108 +614,67 @@ WITH RECURSIVE similar_items AS (
-SELECT * FROM recroute LIMIT 2000;
-----
 
 .Results
```
No need to repeat the results here: you could just have a line saying "The results are the same as for the previous query."

simon-dew reviewed (approved):
LGTM 👍 a couple of notes if you want, but I think this is good to merge as-is.
